File: BaseTest/src/main/java/com/bihe0832/android/base/test/M3U8Info.java
package com.bihe0832.android.base.test;

import com.bihe0832.android.lib.file.FileUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * @author hardyshi code@bihe0832.com
 * Created on 2021/2/20.
 * Description: Description
 */
public class M3U8Info {
    //下载ts的基础URL
    private String baseURL = "";
    //m3u8文件URL
    private String m3u8URL;
    //文件总大小
    private long fileSize;
    //总时间，单位毫秒
    private long totalTime;
    //视频切片
    private List<M3U8TSInfo> tsList = new ArrayList<M3U8TSInfo>();

    public String getBaseURL() {
        return baseURL;
    }

    public void setBaseURL(String baseURL) {
        this.baseURL = baseURL;
    }

    public String getM3u8URL() {
        return m3u8URL;
    }

    public void setM3u8URL(String m3u8URL) {
        this.m3u8URL = m3u8URL;
    }


    public long getFileSize() {
        fileSize = 0;
        for (M3U8TSInfo m3U8Ts : tsList) {
            fileSize = fileSize + m3U8Ts.getFileSize();
        }
        return fileSize;
    }

    public String getFormatFileSize() {
        return FileUtils.INSTANCE.getFileLength(getFileSize());
    }

    public List<M3U8TSInfo> getTsList() {
        return tsList;
    }

    public void addTs(M3U8TSInfo ts) {
        this.tsList.add(ts);
    }

    public long getTotalTime() {
        totalTime = 0;
        for (M3U8TSInfo m3U8Ts : tsList) {
            totalTime = totalTime + (int) (m3U8Ts.getSeconds() * 1000);
        }
        return totalTime;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("m3u8URL: " + m3u8URL);
        sb.append("\nbaseURL: " + baseURL);
        sb.append("\nfileSize: " + getFileSize());
        sb.append("\ntotalTime: " + totalTime);

        for (M3U8TSInfo ts : tsList) {
            sb.append("\nts: " + ts);
        }
        return sb.toString();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof M3U8Info) {
            M3U8Info m3U8 = (M3U8Info) obj;
            return m3u8URL != null && m3u8URL.equals(m3U8.m3u8URL);
        }
        return false;
    }
}

File: BaseM3U8/src/main/java/com/bihe0832/android/base/m3u8/bean/M3U8Info.kt
package com.bihe0832.android.base.m3u8.bean

import com.bihe0832.android.base.m3u8.M3U8ModuleManager
import com.bihe0832.android.lib.file.FileUtils.getFileLength
import java.util.concurrent.CopyOnWriteArrayList

/**
 * @author zixie code@bihe0832.com
 * Created on 2021/2/20.
 * Description: Description
 */
class M3U8Info {

    //Header
    var header = ""

    //下载ts的基础URL
    private var baseURL = ""

    //m3u8文件URL
    private var m3u8URL = ""

    //更新时间
    private var downloadTime: Long = 0

    //文件总大小
    private var fileSize: Long = 0

    //总时间，单位毫秒
    private var totalTime: Long = 0

    //视频切片
    private val tsList: CopyOnWriteArrayList<M3U8TSInfo> = CopyOnWriteArrayList()

    // 实际去重以后得切片数量
    private var mTsSize = 0

    fun getBaseURL(): String {
        return baseURL
    }

    fun setBaseURL(baseURL: String) {
        this.baseURL = baseURL!!
    }

    fun getM3u8URL(): String {
        return m3u8URL
    }

    fun setM3u8URL(m3u8URL: String) {
        this.m3u8URL = m3u8URL
    }

    fun getDownloadTime(): Long {
        return downloadTime
    }

    fun setDownloadTime(downloadTime: Long) {
        this.downloadTime = downloadTime
    }

    fun getFileSize(): Long {
        fileSize = 0
        for (m3U8Ts in tsList) {
            fileSize += m3U8Ts.fileSize
        }
        return fileSize
    }

    fun getFormatFileSize(): String? {
        return getFileLength(getFileSize())
    }

    fun getTsList(): List<M3U8TSInfo> {
        return tsList
    }

    fun getTsSize(): Int {
        if (mTsSize == 0) {
            mTsSize = this.tsList.distinctBy { it.m3u8TSURL }.size
        }
        return mTsSize
    }

    fun addTs(ts: M3U8TSInfo) {
        this.tsList.add(ts)
        this.mTsSize = this.tsList.distinctBy { it.m3u8TSURL }.size
    }

    fun getTotalTime(): Long {
        totalTime = 0
        for (m3U8Ts in tsList) {
            totalTime += (m3U8Ts.seconds * 1000).toInt()
        }
        return totalTime
    }

    fun isOK(): Boolean {
        return tsList.isNotEmpty()
    }

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append("视频信息: \n\n")
        sb.append("m3u8URL: ").append(m3u8URL).append("\n\n")
        sb.append("baseURL: ").append(baseURL).append("\n\n")
        sb.append("保存地址: ").append(M3U8ModuleManager.getDownloadPath(m3u8URL)).append("\n\n")
        sb.append("视频地址: ").append(M3U8ModuleManager.getFinalVideoPath(m3u8URL)).append("\n\n")
        sb.append("分片数量: ").append(tsList.size).append(",").append("实际分片数量: ").append(getTsSize()).append("\n\n")
        if (tsList.isNotEmpty()) {
            sb.append("第一分片信息: ").append(tsList.get(0).toString())
        }
        return sb.toString()
    }


    override fun equals(other: Any?): Boolean {
        if (other is M3U8Info) {
            return m3u8URL != null && m3u8URL == other.m3u8URL
        }
        return false
    }
}